### PR TITLE
feat: add label list index

### PR DESCRIPTION
**Pull request by westonpace**
This PR adds a new kind of scalar index, the LABEL_LIST scalar index.  A label list index can only be created on list columns (the item data type has to be one of the data types that can be used for btree/bitmap indices).  The label list index can speed up array_has_any and array_has_all queries.

This is useful for tag-like columns where each row has some number of labels and the overall cardinality of the labels is relatively low (the index in-memory size will depend on the cardinality of the labels).

This PR also introduces the idea that different scalar indices may support different types of queries.  There is now `SargableQuery` (what is used by btree and bitmap) and `LabelListQuery` (used by the new label list index).  Going forward we should probably break the FullTextQuery out of `SargableQuery` and into its own query type.

### CODE REVIEW

**Comment by wjones127** [comment]
The name collision with version **tags** (#2482) might be unfortunate.

**Comment by chebbyChefNEQ** [comment]
> The name collision with version **tags** (#2482) might be unfortunate.

maybe we could call it `ENUM`?

File: python/python/lance/dataset.py [hunk 1]
```diff
@@ -1213,31 +1219,37 @@ def create_scalar_index(
...
+                and not pa.types.is_boolean(field.type)
+                and not pa.types.is_string(field.type)
+                and not pa.types.is_temporal(field.type)
+            ):
+                raise TypeError(
+                    f"BTREE/BITMAP index column {column} must be int",
+                    ", float, bool, str, or temporal",
+                )
+        elif index_type == "TAG":
+            if not pa.types.is_list(field.type):
```
Comment (chebbyChefNEQ) [inline python/python/lance/dataset.py:1245]: I wonder if we want to support `str` columns. this seems very similar to the idea of "Symbol" in questdb: https://questdb.io/docs/concept/symbol/

File: rust/lance-index/src/scalar/tag.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+        unimplemented!()
+    }
+}
+
+impl TagIndex {
+    async fn search_values(&self, values: &Vec<ScalarValue>) -> Result<Vec<UInt64Array>> {
+        let mut value_results = Vec::with_capacity(values.len());
+        for value in values {
+            let value_query = SargableQuery::Equals(value.clone());
+            value_results.push(self.values_index.search(&value_query).await?);
```
Comment (chebbyChefNEQ) [inline rust/lance-index/src/scalar/tag.rs:85]: tiny nit: could we return iterator here so we don't have to keep the results in memory can can instead union/intersect on the fly

File: python/python/benchmarks/test_search.py [hunk 3]
```diff
@@ -31,16 +31,34 @@ def find_or_clean(dataset_path: Path) -> Union[lance.LanceDataset, None]:
...
+    "documentary",
+]
+
+
 def create_table(num_rows, offset) -> pa.Table:
     values = pc.random(num_rows * N_DIMS).cast(pa.float32())
     vectors = pa.FixedSizeListArray.from_arrays(values, N_DIMS)
     filterable = pa.array(range(offset, offset + num_rows))
     categories = pa.array(np.random.randint(0, 100, num_rows))
+    genres_values = pa.array(np.random.choice(GENRES, num_rows * 3))
```
Comment (chebbyChefNEQ) [inline python/python/benchmarks/test_search.py:53]: Could we also test empty list?

**Review by chebbyChefNEQ — COMMENTED** [review]
barring naming w.r.t. https://github.com/lancedb/lance/pull/2482 lgtm

**Comment by chebbyChefNEQ** [comment]
> > The name collision with version **tags** (#2482) might be unfortunate.
> 
> maybe we could call it `ENUM`?

or maybe should call version tags, version alias instead?

**Comment by westonpace** [comment]
@chebbyChefNEQ @wjones127 how do you two feel about `TAG_LIST`, `LABEL_LIST`, or `ENUM_LIST` for the name?

File: python/python/lance/dataset.py (hunk 1 above)
Comment (westonpace) [reply python/python/lance/dataset.py:1245]: `Symbol` looks like Arrow's `Dictionary` type?

This supports `List<String>`.  I'm not sure `String` (without the `List`) makes sense here.  The functions "array_has_any" and "array_has_all" are not used with a `String` column.  The closest analogue might be storing a bitmap per character or bitmap per word/token but we need a valid use case first (and FTS indices are probably better for that type of query).

Comment (westonpace) [reply python/python/lance/dataset.py:1245]: Also, wrt `Symbol`, another TODO I plan to support this week is support for indices on dictionary columns.  I think the customer in this particular scenario is using `List<Dictionary<String>>` so we eventually want to get there.  `Dictionary<String>` is good for storage purposes but won't have much impact on index performance (unless using btree).

File: python/python/benchmarks/test_search.py [hunk 4]
```diff
@@ -75,6 +93,7 @@ def create_base_dataset(data_dir: Path) -> lance.LanceDataset:
 
     dataset.create_scalar_index("filterable", "BTREE")
     dataset.create_scalar_index("category", "BITMAP")
+    dataset.create_scalar_index("genres", "TAG")
```
Comment (wjones127) [inline python/python/benchmarks/test_search.py:96]: In Postgres, these are called "GIN" or "Generalized Inverted Index". Could call it "INVERTED_STRING_INDEX" or something like that. IMO it's nice to re-use existing names for concepts rather than create new names.

File: python/python/lance/dataset.py [hunk 5]
```diff
@@ -1159,15 +1159,21 @@ def create_scalar_index(
...
-        unique values and many rows per value.
+        The other common index type is ``BITMAP``.  This index stores a bitmap for each
+        unique value in the column.  This index is useful for columns with a small
+        number of unique values and many rows per value.
+
+        The ``TAG`` index type is a special index that is used to index list columns
+        whose values have small cardinality.  For example, a column that contains
+        lists of tags (e.g. ``["tag1", "tag2", "tag3"]``) can be indexed with a
+        ``TAG`` index.  This index can only speedup queries with ``array_has_any``
+        or ``array_has_all`` filters.
```
Comment (wjones127) [inline python/python/lance/dataset.py:1176]: How about `array_contains` / `array_has`?

File: python/python/benchmarks/test_search.py (hunk 4 above)
Comment (wjones127) [reply python/python/benchmarks/test_search.py:96]: Why is the FTS index called inverted index? I wonder if we should move these names around.

File: rust/lance-index/src/scalar/expression.rs [hunk 6]
```diff
@@ -54,6 +58,166 @@ pub struct IndexedExpression {
...
+            return None;
+        }
+        let tag_list = maybe_scalar(&args[1], data_type)?;
+        if let ScalarValue::List(list_arr) = tag_list {
+            let list_values = list_arr.values();
+            let mut scalars = Vec::with_capacity(list_values.len());
+            for idx in 0..list_values.len() {
+                scalars.push(ScalarValue::try_from_array(list_values.as_ref(), idx).ok()?);
+            }
+            if func.name() == "array_has_all" {
```
Comment (wjones127) [inline rust/lance-index/src/scalar/expression.rs:200]: IMO we should also handle `array_has`. (`array_contains` is an alias for this, I believe). https://datafusion.apache.org/user-guide/sql/scalar_functions.html#array-contains

File: python/python/benchmarks/test_search.py (hunk 4 above)
Comment (wjones127) [reply python/python/benchmarks/test_search.py:96]: I'm fine with label list index.

Comment (westonpace) [reply python/python/benchmarks/test_search.py:96]: > Why is the FTS index called inverted index? I wonder if we should move these names around.

"inverted index" is a very general concept.  Normally you have a structure that goes from row id to value.  "Inverted index" just menas you have a structure that goes from value to row id.  Pretty much every index is an inverted index.  Inverted indices are often used for full text search.  E.g. TF-IDF and BM25 are both "inverted indices".

IVF is an inverted index.

Also, GIN, in postgres, is much more general (e.g. it can be used for full text search on a string column) and I'm not sure a GIN index on an Array column is doing the same thing as we are doing here.

File: rust/lance-index/src/scalar/expression.rs (hunk 6 above)
Comment (westonpace) [reply rust/lance-index/src/scalar/expression.rs:200]: Good point, we can map `array_has` to `array_has_any`

File: python/python/benchmarks/test_search.py (hunk 4 above)
Comment (westonpace) [reply python/python/benchmarks/test_search.py]: Ok, yeah, reading more, I suppose one could think of this as a GIN index.

File: rust/lance-index/src/scalar/tag.rs (hunk 2 above)
Comment (westonpace) [reply rust/lance-index/src/scalar/tag.rs]: It's a bit trickier since the `values_index.search` method is async but I made it work with streams.

File: python/python/benchmarks/test_search.py (hunk 3 above)
Comment (westonpace) [reply python/python/benchmarks/test_search.py:53]: Empty list is tested in the rust test (it uses the random datagen utilities and I verified they generate empty lists)

**Comment by westonpace** [comment]
I fixed a few bugs and addressed some of the review feedback.

**Review by wjones127 — APPROVED** [review]
